cirq-pasqal/cirq_pasqal/pasqal_sampler.py

```diff
@@ -114,8 +133,13 @@ def run_sweep(
             Result list for this run; one for each possible parameter
             resolver.
         """
-        assert isinstance(program.device, cirq_pasqal.PasqalDevice)
-        program.device.validate_circuit(program)
+        if program._device != cirq.UNCONSTRAINED_DEVICE:
+            assert isinstance(
+                program._device, cirq_pasqal.PasqalDevice
+            ), "Device must inherit from cirq.PasqalDevice."
+            program._device.validate_circuit(program)
+        elif self._device:
+            self._device.validate_circuit(program)
```

Comment:
How about this? 
```suggestion
        device = (
            program._device if program._device != cirq.UNCONSTRAINED_DEVICE 
            else self._device
            )
        assert isinstance(
            device, cirq_pasqal.PasqalDevice
            ), "Device must inherit from cirq.PasqalDevice."
        device.validate_circuit(program)
```

Reply:
Ahhh I see. The major difference between what is implemented and this is:

1. It will type check the new self._device.
2. It will no longer allow a user to do no device verification at all.

This seems fine to me. I will say that generally speaking it's not always a great practice to type check inputs for their expected type in python. This is because now your function boundary which was defined clearly on the types it accepted is now defined on all possible types and works correctly on the desired input types, where it doesn't work is now obscured by this type check that doesn't allow things to crash and provide a full trace. Another way of thinking about this is: this would paradoxically make behavior under violation of the API part of the API. 

I'll implement this new way now and if anyone feels strongly we can always revisit in a later PR.